### ISSUE TITLE
Not possible to update individual 'po' files

### ISSUE THREAD

**Issue by stephenfin** [sphinx-doc/sphinx#4476]
### Problem
`sphinx-build` allows you to pass a list of filenames, which should ensure that only these files are rebuilt. This appears to work as expected. However, if you also have i18n features enabled, then it seems you should be able to build message catalogs at the same time, as seen [here](https://github.com/sphinx-doc/sphinx/blob/1.7.0b1/sphinx/application.py#L327). Unfortunately, this appears to be broken by design.

#### Procedure to reproduce the problem
1. Modify `conf.py` to enable English language "translations".

    ```
    diff --git a/doc/conf.py b/doc/conf.py
    index fa82cbfb7..5f2d66e90 100644
    --- a/doc/conf.py
    +++ b/doc/conf.py
    @@ -100,6 +100,7 @@ intersphinx_mapping = {'python': ('https://docs.python.org/2/', None)}
 
     # Sphinx document translation with sphinx gettext feature uses these settings:
     locale_dirs = ['locale/']
    +language = 'en'
     gettext_compact = False
    ```

2. Generate message catalogs, as described in the docs.

    ```
    pip install sphinx-intl
    cd docs
    make gettext
    sphinx-intl update -p _build/gettext -l de -l ja
    ```

3. Build entire docs once then build a single file.

    ```
    sphinx-build doc/ -d build/sphinx/doctrees/ build/sphinx/html/
    sphinx-build doc/ -d build/sphinx/doctrees/ build/sphinx/html/ doc/latex.rst
    ```

#### Error logs / results
```
$ rm doc/locale/en/LC_MESSAGES/*.mo doc/locale/en/LC_MESSAGES/*/*.mo

$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/
Running Sphinx v1.8.0+/9a3f401c4
loading translations [en]... done
building [mo]: targets for 70 po files that are out of date
writing output... [100%] locale/en/LC_MESSAGES/extdev/nodes.mo
building [html]: targets for 70 source files that are out of date
updating environment: 70 added, 0 changed, 0 removed
reading sources... [100%] websupport
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] websupport
generating indices... genindex py-modindex
highlighting module code... [100%] sphinxcontrib.websupport.storage
writing additional pages... index search opensearch
copying images... [100%] themes/alabaster.png
copying downloadable files... [100%] ext/example_google.py
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.

The HTML pages are in build/sphinx/html.

$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/ doc/ext/math.rst
Running Sphinx v1.8.0+/9a3f401c4
loading translations [en]... done
building [mo]: targets for 0 po files that are specified
building [html]: 1 source files given on command line
updating environment: 70 added, 0 changed, 0 removed
reading sources... [100%] websupport
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] ext/math
generating indices... genindex py-modindex
highlighting module code... [100%] sphinxcontrib.websupport.storage
writing additional pages... index search opensearch
copying downloadable files... [100%] ext/example_google.py
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.

The HTML pages are in build/sphinx/html.
```

The particular issue is this line from the second build:

```
$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/ doc/ext/math.rst
...
building [mo]: targets for 0 po files that are specified
```

I would expect something similar to the below:

```
$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/ doc/ext/math.rst
...
building [mo]: targets for 1 po files that are specified
writing output... [100%] locale/en/LC_MESSAGES/ext/math.mo
```

#### Expected results
Assuming I understand the feature correctly, the catalog for a given file should be updated regardless of whether it's supposed to be out-of-date and nothing else should be. Alternatively, perhaps this file should not be updated unless it's out of date, in which case, how should one use this?

### Reproducible project / your project
- Sphinx docs

### Environment info
- OS: Fedora 27 (Linux)
- Python version: 2.7 (tox)
- Sphinx version: master (9a3f401c469ec6c16be32ce1bda11d266ba88d76)

**Comment by stephenfin** [comment]
I applied the following hacky change to try and see what's going on.

```diff
diff --git a/sphinx/util/i18n.py b/sphinx/util/i18n.py
index 75a8506fa..27f46964a 100644
--- a/sphinx/util/i18n.py
+++ b/sphinx/util/i18n.py
@@ -121,6 +121,16 @@ def find_catalog_source_files(locale_dirs, locale, domains=None, gettext_compact
     """
     catalogs = set()  # type: Set[CatalogInfo]
 
+    print('!! domains')
+    print(domains)
+    print('!! locale')
+    print(locale)
+
+    domains = domains or []
+    domains = [d.split('/', 1)[1] if d.startswith('doc/') else d for d in domains]
+    print('!!! domains v2')
+    print(domains)
+
     if not locale:
         return catalogs  # locale is not specified
 
@@ -128,6 +138,11 @@ def find_catalog_source_files(locale_dirs, locale, domains=None, gettext_compact
         if not locale_dir:
             continue  # skip system locale directory
 
+        print('!! locale_dir')
+        print(locale_dir)
+
+        found_domains = []
+
         base_dir = path.join(locale_dir, locale, 'LC_MESSAGES')
 
         if not path.exists(base_dir):
@@ -141,12 +156,20 @@ def find_catalog_source_files(locale_dirs, locale, domains=None, gettext_compact
                 if gettext_compact and path.sep in domain:
                     domain = path.split(domain)[0]
                 domain = domain.replace(path.sep, SEP)
+                found_domains.append(domain)
                 if domains and domain not in domains:
                     continue
                 cat = CatalogInfo(base_dir, domain, charset)
+                print('!!! catlog')
+                print(cat)
+                print('force_all? %r\tis_outdated? %r' % (
+                    bool(force_all), bool(cat.is_outdated())))
                 if force_all or cat.is_outdated():
                     catalogs.add(cat)
 
+        print('!!! found domains')
+        print(found_domains)
+
     return catalogs
```

This yields the following output.

```
$ rm doc/locale/en/LC_MESSAGES/*.mo doc/locale/en/LC_MESSAGES/*/*.mo

$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/                                                                                                                                                Running Sphinx v1.8.0+/9a3f401c4                                                                                                                                                                                   loading translations [en]... !! domains                                                                                                                                                                            ['sphinx']
!! locale
en
!!! domains v2
['sphinx']
!! locale_dir
/home/sfinucan/Development/sphinx/sphinx/doc/locale/
Catlog
CatalogInfo(base_dir=u'/home/sfinucan/Development/sphinx/sphinx/doc/locale/en/LC_MESSAGES', domain=u'sphinx', charset='utf-8-sig')
force_all? False        is_outdated? True
!!! found domains
[u'latex', u'install', u'templating', u'faq', u'authors', u'domains', u'changes', u'tutorial', u'intro', u'config', u'contents', u'develop', u'setuptools', u'sphinx', u'devguide', u'theming', u'glossary', u'rest
', u'builders', u'websupport', u'markdown', u'examples', u'extensions', u'intl', u'web/api', u'web/storagebackends', u'web/quickstart', u'web/searchadapters', u'extdev/markupapi', u'extdev/domainapi', u'extdev/a
ppapi', u'extdev/parserapi', u'extdev/envapi', u'extdev/logging', u'extdev/builderapi', u'extdev/collectorapi', u'extdev/tutorial', u'extdev/index', u'extdev/nodes', u'ext/graphviz', u'ext/napoleon', u'ext/autos
ummary', u'ext/coverage', u'ext/example_google', u'ext/thirdparty', u'ext/inheritance', u'ext/extlinks', u'ext/builtins', u'ext/linkcode', u'ext/math', u'ext/viewcode', u'ext/ifconfig', u'ext/todo', u'ext/imgcon
verter', u'ext/autosectionlabel', u'ext/doctest', u'ext/githubpages', u'ext/intersphinx', u'ext/example_numpy', u'ext/autodoc', u'markup/misc', u'markup/toctree', u'markup/para', u'markup/code', u'markup/index',
 u'markup/inline', u'man/sphinx-autogen', u'man/sphinx-quickstart', u'man/sphinx-build', u'man/index', u'man/sphinx-apidoc']
done
!! domains
None
!! locale
en
!!! domains v2
[]
!! locale_dir
/home/sfinucan/Development/sphinx/sphinx/doc/locale/
!!! catlog
CatalogInfo(base_dir=u'/home/sfinucan/Development/sphinx/sphinx/doc/locale/en/LC_MESSAGES', domain=u'latex', charset='utf-8-sig')
force_all? False        is_outdated? True

   ... [snip] ...

!!! catlog
CatalogInfo(base_dir=u'/home/sfinucan/Development/sphinx/sphinx/doc/locale/en/LC_MESSAGES', domain=u'man/sphinx-apidoc', charset='utf-8-sig')
force_all? False        is_outdated? True
!!! found domains
[u'latex', u'install', u'templating', u'faq', u'authors', u'domains', u'changes', u'tutorial', u'intro', u'config', u'contents', u'develop', u'setuptools', u'sphinx', u'devguide', u'theming', u'glossary', u'rest', u'builders', u'websupport', u'markdown', u'examples', u'extensions', u'intl', u'web/api', u'web/storagebackends', u'web/quickstart', u'web/searchadapters', u'extdev/markupapi', u'extdev/domainapi', u'extdev/appapi', u'extdev/parserapi', u'extdev/envapi', u'extdev/logging', u'extdev/builderapi', u'extdev/collectorapi', u'extdev/tutorial', u'extdev/index', u'extdev/nodes', u'ext/graphviz', u'ext/napoleon', u'ext/autosummary', u'ext/coverage', u'ext/example_google', u'ext/thirdparty', u'ext/inheritance', u'ext/extlinks', u'ext/builtins', u'ext/linkcode', u'ext/math', u'ext/viewcode', u'ext/ifconfig', u'ext/todo', u'ext/imgconverter', u'ext/autosectionlabel', u'ext/doctest', u'ext/githubpages', u'ext/intersphinx', u'ext/example_numpy', u'ext/autodoc', u'markup/misc', u'markup/toctree', u'markup/para', u'markup/code', u'markup/index', u'markup/inline', u'man/sphinx-autogen', u'man/sphinx-quickstart', u'man/sphinx-build', u'man/index', u'man/sphinx-apidoc']
building [mo]: targets for 70 po files that are out of date
writing output... [100%] locale/en/LC_MESSAGES/extdev/nodes.mo
building [html]: targets for 70 source files that are out of date
updating environment: 70 added, 0 changed, 0 removed
reading sources... [100%] websupport
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] websupport
generating indices... genindex py-modindex
highlighting module code... [100%] sphinxcontrib.websupport.storage
writing additional pages... index search opensearch
copying images... [100%] themes/alabaster.png
copying downloadable files... [100%] ext/example_google.py
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.

The HTML pages are in build/sphinx/html.

$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/ doc/ext/math.rst
Running Sphinx v1.8.0+/9a3f401c4
loading translations [en]... !! domains
['sphinx']
!! locale
en
!!! domains v2
['sphinx']
!! locale_dir
/home/sfinucan/Development/sphinx/sphinx/doc/locale/
!!! catlog
CatalogInfo(base_dir=u'/home/sfinucan/Development/sphinx/sphinx/doc/locale/en/LC_MESSAGES', domain=u'sphinx', charset='utf-8-sig')
force_all? False        is_outdated? False
!!! found domains
[u'latex', u'install', u'templating', u'faq', u'authors', u'domains', u'changes', u'tutorial', u'intro', u'config', u'contents', u'develop', u'setuptools', u'sphinx', u'devguide', u'theming', u'glossary', u'rest
', u'builders', u'websupport', u'markdown', u'examples', u'extensions', u'intl', u'web/api', u'web/storagebackends', u'web/quickstart', u'web/searchadapters', u'extdev/markupapi', u'extdev/domainapi', u'extdev/a
ppapi', u'extdev/parserapi', u'extdev/envapi', u'extdev/logging', u'extdev/builderapi', u'extdev/collectorapi', u'extdev/tutorial', u'extdev/index', u'extdev/nodes', u'ext/graphviz', u'ext/napoleon', u'ext/autos
ummary', u'ext/coverage', u'ext/example_google', u'ext/thirdparty', u'ext/inheritance', u'ext/extlinks', u'ext/builtins', u'ext/linkcode', u'ext/math', u'ext/viewcode', u'ext/ifconfig', u'ext/todo', u'ext/imgcon
verter', u'ext/autosectionlabel', u'ext/doctest', u'ext/githubpages', u'ext/intersphinx', u'ext/example_numpy', u'ext/autodoc', u'markup/misc', u'markup/toctree', u'markup/para', u'markup/code', u'markup/index',
 u'markup/inline', u'man/sphinx-autogen', u'man/sphinx-quickstart', u'man/sphinx-build', u'man/index', u'man/sphinx-apidoc']
done
!! domains
['doc/ext/math']
!! locale
en
!!! domains v2
['ext/math']
!! locale_dir
/home/sfinucan/Development/sphinx/sphinx/doc/locale/
!!! catlog
CatalogInfo(base_dir=u'/home/sfinucan/Development/sphinx/sphinx/doc/locale/en/LC_MESSAGES', domain=u'ext/math', charset='utf-8-sig')
force_all? False        is_outdated? False
!!! found domains
[u'latex', u'install', u'templating', u'faq', u'authors', u'domains', u'changes', u'tutorial', u'intro', u'config', u'contents', u'develop', u'setuptools', u'sphinx', u'devguide', u'theming', u'glossary', u'rest
', u'builders', u'websupport', u'markdown', u'examples', u'extensions', u'intl', u'web/api', u'web/storagebackends', u'web/quickstart', u'web/searchadapters', u'extdev/markupapi', u'extdev/domainapi', u'extdev/a
ppapi', u'extdev/parserapi', u'extdev/envapi', u'extdev/logging', u'extdev/builderapi', u'extdev/collectorapi', u'extdev/tutorial', u'extdev/index', u'extdev/nodes', u'ext/graphviz', u'ext/napoleon', u'ext/autos
ummary', u'ext/coverage', u'ext/example_google', u'ext/thirdparty', u'ext/inheritance', u'ext/extlinks', u'ext/builtins', u'ext/linkcode', u'ext/math', u'ext/viewcode', u'ext/ifconfig', u'ext/todo', u'ext/imgcon
verter', u'ext/autosectionlabel', u'ext/doctest', u'ext/githubpages', u'ext/intersphinx', u'ext/example_numpy', u'ext/autodoc', u'markup/misc', u'markup/toctree', u'markup/para', u'markup/code', u'markup/index',
 u'markup/inline', u'man/sphinx-autogen', u'man/sphinx-quickstart', u'man/sphinx-build', u'man/index', u'man/sphinx-apidoc']
building [mo]: targets for 0 po files that are specified
building [html]: 1 source files given on command line
updating environment: 70 added, 0 changed, 0 removed
reading sources... [100%] websupport
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] ext/math
generating indices... genindex py-modindex
highlighting module code... [100%] sphinxcontrib.websupport.storage
writing additional pages... index search opensearch
copying downloadable files... [100%] ext/example_google.py
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.

The HTML pages are in build/sphinx/html
```

There are two things to note here.

1. I had to strip the `doc/` prefix from the provided path, because otherwise the domains would never match. This appears to be a bug.

    ```diff
    +    domains = domains or []
    +    domains = [d.split('/', 1)[1] if d.startswith('doc/') else d for d in domains]
    +    print('!!! domains v2')
    +    print(domains)
    ```

2. Even with that patch applied, I still didn't see the files rebuilt.

    ```
    !!! catlog
    CatalogInfo(base_dir=u'/home/sfinucan/Development/sphinx/sphinx/doc/locale/en/LC_MESSAGES', domain=u'ext/math', charset='utf-8-sig')
    force_all? False        is_outdated? False
    ```

    This is because neither of the pre-conditions for rebuilding is true. However, this seems counter-intuitive: surely we should force the rebuild in this case?

**Comment by stephenfin** [comment]
Final comments. We seem to be running translation twice: once during the `sphinx.application.Application._init_i18n` function, and another time during the `sphinx.application.Application.build` function (via a call to `sphinx.builders.Builder.compile_specific_catalogs`). I'm not sure I really get the purpose of the former. Some improved docstrings here would be much appreciated.

**Comment by tk0miya** [comment]
>1. I had to strip the doc/ prefix from the provided path, because otherwise the domains would never match. This appears to be a bug.

>2. Even with that patch applied, I still didn't see the files rebuilt.

I think 2. is not a bug. .po files are not target of rebuild.
Is there reason to rebuild them again? The compilation of .po files is not related with Sphinx documentation directly. So I don't think the rebuild is not needed.

But 1. might be a bug. If .mo files not exists, we have to compile them before building docs.

**Comment by stephenfin** [comment]
> I think 2. is not a bug. .po files are not target of rebuild.
> Is there reason to rebuild them again? The compilation of .po files is not related with Sphinx documentation directly. So I don't think the rebuild is not needed.

I could be using the wrong terminology here (I've never worked with Sphinx's translation machinery before and the documentation assumes a lot of pre-requisite knowledge which I don't have). `po` files are generated by `gettext` and modified by a user and then `.mo` files are written by Sphinx...right? What I'm referring to are these lines.

```python
class Sphinx(object):

    # ...

    def build(self, force_all=False, filenames=None):
        # type: (bool, List[unicode]) -> None
        try:
            if force_all:
                self.builder.compile_all_catalogs()
                self.builder.build_all()
            elif filenames:
                self.builder.compile_specific_catalogs(filenames)  # <-- *** here ***
                self.builder.build_specific(filenames)
            else:
                # ....
```

From [here](https://github.com/sphinx-doc/sphinx/blob/1.6.6/sphinx/application.py#L334-L336).

Surely this suggest we should be re-compiling the `.mo` files? If not, why does that function exist?

**Comment by tk0miya** [comment]
Yes, your understanding is right.

>Surely this suggest we should be re-compiling the .mo files? If not, why does that function exist?

In my understanding, it is used for first compilation, not re-compilation.
Certainly, this compiles .po files to .mo files. I think it is useful for document who doesn't have `.mo` files on there repository. For example, document of Sphinx also doesn't have `.mo` files on repository.
On the other hand, the `.mo` files is not changed even if Sphinx rebuilds docs. Re-compilation is only needed if `.po` file is updated.

**Comment by tk0miya** [comment]
ping @stephenfin 
Is this still needed?

**Comment by stephenfin** [comment]
I need to check again but I don't think this has been fixed, no

**Comment by stephenfin** [comment]
Looks like this was resolved by the fix for #4560.
